pkg/controller/pxc/status.go

```diff
@@ -190,14 +147,59 @@ func (r *ReconcilePerconaXtraDBCluster) updateStatus(cr *api.PerconaXtraDBCluste
 		if cr.Status.ProxySQL.Message != "" {
 			cr.Status.Messages = append(cr.Status.Messages, "ProxySQL: "+cr.Status.ProxySQL.Message)
 		}
-		inProgres, err = r.upgradeInProgress(cr, "proxysql")
+		*inProgres, err = r.upgradeInProgress(cr, "proxysql")
 		if err != nil {
 			return fmt.Errorf("check proxysql upgrade progress: %v", err)
 		}
 	} else {
 		cr.Status.ProxySQL = api.AppStatus{}
 	}
 
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) updateStatus(cr *api.PerconaXtraDBCluster, reconcileErr error) (err error) {
+	clusterCondition := api.ClusterCondition{
+		Status:             api.ConditionTrue,
+		Type:               api.ClusterInit,
+		LastTransitionTime: metav1.NewTime(time.Now()),
+	}
+
+	if reconcileErr != nil {
+		if cr.Status.Status != api.ClusterError {
+			clusterCondition = api.ClusterCondition{
+				Status:             api.ConditionTrue,
+				Type:               api.ClusterError,
+				Message:            reconcileErr.Error(),
+				Reason:             "ErrorReconcile",
+				LastTransitionTime: metav1.NewTime(time.Now()),
+			}
+			cr.Status.Conditions = append(cr.Status.Conditions, clusterCondition)
+
+			cr.Status.Messages = append(cr.Status.Messages, "Error: "+reconcileErr.Error())
+			cr.Status.Status = api.ClusterError
+		}
+
+		return r.writeStatus(cr)
+	}
+
+	cr.Status.Messages = cr.Status.Messages[:0]
+	clusterStatus := clusterStatus{}
+
+	if err := r.updatePXCStatus(cr, &clusterStatus); err != nil {
+		return err
+	}
+
+	inProgres := false
+
+	if err := r.updateHAProxyStatus(cr, &inProgres, &clusterStatus); err != nil {
```

Comment:
I don't quite like an idea that communication is built around changing the parameters passed to the method.
Now, these methods do a bit chaotic meaning that the do a lot of different stuff. Plus these set a lot of states. It's too unclear.